src/NServiceBus.Core/ObjectBuilder/ContainerComponent.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Collections.Generic;
+using NServiceBus.Container;
+using NServiceBus.ObjectBuilder;
+using NServiceBus.ObjectBuilder.Common;
+using NServiceBus.Settings;
+
+namespace NServiceBus
+{
+    class ContainerComponent
+    {
+        public ContainerComponent(SettingsHolder settings)
+        {
+            this.settings = settings;
+        }
+
+        public void UseContainer<T>(Action<ContainerCustomizations> customizations = null) where T : ContainerDefinition, new()
+        {
+            customizations?.Invoke(new ContainerCustomizations(settings));
+
+            UseContainer(typeof(T));
+        }
+
+        public void UseContainer(Type definitionType)
+        {
+            UseContainer(definitionType.Construct<ContainerDefinition>().CreateContainer(settings));
+        }
+
+        public void UseExternallyManagedContainer(IConfigureComponents configureComponents)
+        {
+            if (customContainer != null)
+            {
+                throw new InvalidOperationException("Configuring an explicit internal container via the `EndpointConfiguration.UseContainer` API is not supported when using an external container and can safely be removed.");
```

Comment:
I'm not sure I fully understand the exception scenario, but I think it should first state the _immediate_ problem, i.e. that the consumer is trying to configure an external container when an internal container has already been configured, and then explain why that is a problem.

```suggestion
                throw new InvalidOperationException("An explicit internal container has already been configured using `EndpointConfiguration.UseContainer`. It is not possible to use both an explicit internal container and an externally managed container.");
```

BTW - what about the reverse case, when someone configures an external container and _then_ configures an internal container?

Reply:
The API that requires you to call Prepare and then the Start overload that accepts a "PreparedEndpoint" makes that impossible?